### ISSUE TITLE
[BUG] StringFormatter `nan_format` seems broken 

### ISSUE THREAD

**Issue by hoxbro** [bokeh/bokeh#14014]
### Software versions

``` yaml
Python version        :  3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:23:07) [GCC 12.3.0]
IPython version       :  8.26.0
Tornado version       :  6.4.1
Bokeh version         :  3.5.1
BokehJS static path   :  /home/shh/miniconda3/envs/holoviz/lib/python3.12/site-packages/bokeh/server/static
node.js version       :  v22.5.1
npm version           :  10.8.2
jupyter_bokeh version :  4.0.5
Operating system      :  Linux-6.9.3-76060903-generic-x86_64-with-glibc2.35
```

### Expected behavior

I would expect `NaN` values to be changed to `-`.

### Observed behavior

It does not. 

### Example code

```Python
import numpy as np
from bokeh.io import show
from bokeh.models import ColumnDataSource, DataTable, StringFormatter, TableColumn

formatter = StringFormatter(nan_format="-")
source = ColumnDataSource(dict(a=["1", "2", "3", np.nan, float("nan")]))
columns = [TableColumn(field="a", formatter=formatter)]
data_table = DataTable(source=source, columns=columns)
show(data_table)
```

### Screenshots

![image](https://github.com/user-attachments/assets/3695e26b-bb3e-468f-8f24-fc7118a831cb)

**Comment by bryevdv** [comment]
Well the proximate cause is simple, base `StringFormatter` defines `nan_format` but does not use or refer to it at all in `doFormat`:

https://github.com/bokeh/bokeh/blob/61b87ac7e47df2d920635b4f13a89e00062f07f1/bokehjs/src/lib/models/widgets/tables/cell_formatters.ts#L54-L156

@philippjfr implemented #10416 can you comment? I am not sure why `nan_format` should be an option on `StringFormatter` to be honest. But if we keep it there then `doFormat` needs an update.

**Comment by philippjfr** [comment]
Seems like my original PR didn't introduce it on `StringFormatter`, and it was moved up the class hierarchy in https://github.com/bokeh/bokeh/pull/12098. Not strictly sure it was necessary in that PR either, but I guess I'd be okay with implementing it for `StringFormatter` if it's already there (and removing it would require a deprecation cycle despite it never being functional).
